caf: ble_bond: Introduce CAF BLE bond implementation (#6525)
nrfconnect/sdk-nrf: doc/nrf/libraries/caf/ble_bond.rst

.. _caf_ble_bond:

CAF: Bluetooth LE bond module
#############################

.. contents::
   :local:
   :depth: 2

The |ble_bond| of the :ref:`lib_caf` (CAF) allows to introduce Bluetooth LE bonding to the application.
The module also allows to delete bonds for the default Bluetooth local identity.

Configuration
*************

Before using the module make sure that the following Kconfig options are enabled:

* :kconfig:`CONFIG_BT_BONDABLE`
* :kconfig:`CONFIG_BT_SETTINGS`
* :kconfig:`CONFIG_CAF_SETTINGS_LOADER`
* :kconfig:`CONFIG_CAF_BLE_COMMON_EVENTS`

> The device must be Bluetooth bondable and bonds must be stored in non-volatile memory (settings).

Reviewer: ```suggestion
The device must require Bluetooth bonding and bonds must be stored in the non-volatile memory (settings).
```

Author: Device must be bondable. It does not have to require bonding actually.

Author: Added missing `the`

The :ref:`caf_settings_loader` is used to load content of the settings area during boot.

> Enable the |ble_bond| module using :kconfig:`CONFIG_CAF_BLE_BOND`.

Reviewer: ```suggestion
The |ble_bond| module is enabled using :kconfig:`CONFIG_CAF_BLE_BOND`.
```

Author: Should we use active or passive form here? @greg-fer preferred active form if I recall correctly

Reviewer: Yes, we should be using active tense, but when we start a new paragraph with a new sentence, it’s good to start with a noun.

Author: Fixed

> This Kconfig option selects :kconfig:`CONFIG_CAF_BLE_BOND_SUPPORTED` option to indicate that Bluetooth LE bond functionality is supported by the application.

Reviewer: ```suggestion
This Kconfig option selects :kconfig:`CONFIG_CAF_BLE_BOND_SUPPORTED` option to indicate that the |ble_bond| functionality is supported by the application.
```

Author: Added missing `the`. We cannot use `|ble_bond|` here, because it expands to `Bluetooth LE bond module`

Author: Rephrased this one a bit.

Enabling bond erase
===================

> The |ble_bond| can erase bonds for the default Bluetooth local identity when predefined button click is detected.

Reviewer: ```suggestion
The |ble_bond| can erase bonds for the default Bluetooth local identity when a predefined button click is detected.
```

Author: Added

> The module subscribes for :c:struct:`click_event` to detect button clicks.

Reviewer: ```suggestion
The module subscribes for :c:struct:`click_event` to detect the button clicks.
```

Complete the following steps to enable bond erase:

> 1. Set the :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK` to enable the bond erase when predefined button click is detected.

Reviewer: ```suggestion
1. Set the :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK` to enable the bond erase when a predefined button click is detected.
```

#. Set :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK_KEY_ID` to define the key ID of the button used to erase bonds.

> #. Select one of the following Kconfig options to specify the click type (:c:enum:`click`) that triggers bond erase:

Reviewer: ```suggestion
#. Select one of the following Kconfig options to specify the click type (:c:enum:`click`) that triggers the bond erase:
```

   * :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK_NONE` - :c:enumerator:`CLICK_NONE`,
   * :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK_SHORT` - :c:enumerator:`CLICK_SHORT`,
   * :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK_LONG` - :c:enumerator:`CLICK_LONG`,
   * :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK_DOUBLE` - :c:enumerator:`CLICK_DOUBLE`.

> #. By default, detection of the specified click for specified button always triggers the bond erase.

Reviewer: ```suggestion
#. By default, detection of the specific click for a specific button always triggers the bond erase.
```

>    Set :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK_TIMEOUT` to specify the time of waiting for the button click detection after boot.

Reviewer: ```suggestion
   Set :kconfig:`CONFIG_CAF_BLE_BOND_PEER_ERASE_CLICK_TIMEOUT` to specify the waiting time for detecting the button click after boot.
```

   The timeout is specified in milliseconds.
   The button click is ignored if it occurrs after the timeout.

Implementation details
**********************

The |ble_bond| can be used as a default implementation of Bluetooth LE bond functionality for simple applications.
The module does not broadcast information about performed Bluetooth LE peer operations using :c:struct:`ble_peer_operation_event`.
The module assumes that only default Bluetooth local identity is used.

.. note::

>    If your applications requires an application-specific Bluetooth LE bond and peer management, you must provide your own implementation of the Bluetooth LE bond module.

Reviewer: ```suggestion
   If your application requires an application-specific Bluetooth LE bond and peer management, you must provide your own implementation of the Bluetooth LE bond module.
```
